ow_power_system/src/PowerSystemNode.cpp

```diff
@@ -4,194 +4,807 @@
+    builder.setConfigParam("Predictor.Horizon", std::to_string(m_max_horizon_secs));
+
+    // Create the prognosers using the builder that will send predictions using
+    // their corresponding message bus.
+    for (int i = 0; i < NUM_MODELS; i++)
+    {
+      prognosers.push_back(builder.build(m_power_models[i].bus,
+                                         m_power_models[i].name,
+                                         "trajectory"));
+    }
+  }
+  catch(const std::exception& err)
+  {
+    ROS_ERROR_STREAM("OW_POWER_SYSTEM ERROR: " << err.what()
+                     << " while attempting to initialize asynchronous "
+                     << "prognosers!\nEnsure /config/async_prognoser.cfg exists "
+                     << "and is formatted properly before restarting the "
+                     << "simulation.");
+    return;
   }
 
-  m_init_time = system_clock::now(); // Taken from initPrognoser()
+  ROS_INFO_STREAM("Power system node running.");
+
+  // This rate object is used to sync the cycles up to the provided Hz.
+  // NOTE: If the cycle takes longer than the provided Hz, nothing bad will
+  //       necessarily occur, but the simulation will be out of sync with real
+  //       time.
+  ros::Rate rate(m_gsap_rate_hz);
+
+  // Loop through the PowerModelHandlers to update their values and send them to
+  // the bus each loop to trigger predictions.
+  // This loop should run at 0.5Hz, publishing predictions every 2 seconds via rostopics.
+  int start_loops = NUM_MODELS;
+  while(ros::ok())
+  {
+    ros::spinOnce();
+
+    // Set up value modifiers in each model handler for injection later.
+    injectFaults();
 
+    // DEBUG PRINT
+    if (m_timestamp_print_debug)
+    {
+      printTimestamp(m_power_models[0].model_info.timestamp);
+    }
+
+    double excessive_draw = 0.0;
+
+    // Cycle the power stats in the models.
+    for (int i = 0; i < NUM_MODELS; i++)
+    {
+      double temp_draw = m_power_models[i].model.RunOnce();
+      // Signal the warning message to be displayed after the loop.
+      if (temp_draw > 0)
+      {
+        // Current implementation means if one input exceeds the draw limit,
+        // all inputs will exceed at the same amount, so it doesn't matter which
+        // cell input we take.
+        excessive_draw = temp_draw;
+      }
+
+      m_power_models[i].model.GetPowerStats(m_power_models[i].model_info);
+      
+      // If the timestamp is the same as the previous one (happens during startup),
+      // do not publish the data to prevent crashes.
+      if (m_power_models[i].previous_time != m_power_models[i].model_info.timestamp)
+      {
+        auto timestamp = START_TIME + std::chrono::milliseconds(
+          static_cast<unsigned>(m_power_models[i].model_info.timestamp * 1000)
+        );
+        
+        // Compile a vector<shared_ptr<DoubleMessage>> and then individually
+        // publish each component.
+        std::vector<std::shared_ptr<DoubleMessage>> data_to_publish;
```

Comment:
```suggestion
```
This variable looks to be no longer in use.

Reply:
Is this issue resolved?

Reply:
Yup, I removed it. Forgot to resolve this conversation as a result, whoops.